### PR TITLE
Update CEMS partitions to handle year-quarter files

### PR DESCRIPTION
**Pull request by e-belfer**
Make CEMS quarterly and add 2023 data! See #2973 for detailed task list.

This PR:
- Updates `pudl.extract.epacems` to read in quarterly data
- Updates `pudl.transform.epacems` to handle quarterly data
- Updates `pudl.etl.epacems_assets` to write year/state row groups from quarterly parquet files to the monolithic output
- Reduces dagster concurrency of `hourly_emissions_epacems.process_single_year()` to prevent OOM issues (currently to 2 threads)
- Updates integration tests to handle new EPA CEMS format

# PR Checklist

- [x] Merge the most recent version of the branch you are merging into (probably `dev`).
- [x] All CI checks are passing. [Run tests locally to debug failures](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#running-tests-with-tox)
- [x] Make sure you've included good docstrings.
- [x] For major data coverage & analysis changes, [run data validation tests](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#data-validation)
- [x] Update the [release notes](https://catalystcoop-pudl.readthedocs.io/en/latest/release_notes.html) and reference reference the PR and related issues.
- [x] Do your own explanatory review of the PR to help the reviewer understand what's going on and identify issues preemptively.

### CODE REVIEW

File: src/pudl/workspace/datastore.py [hunk 1]
```diff
@@ -181,8 +181,8 @@ class ZenodoDoiSettings(BaseSettings):
     # eia_bulk_elec: ZenodoDoi = "10.5072/zenodo.1103572"
     epacamd_eia: ZenodoDoi = "10.5281/zenodo.7900974"
     # epacamd_eia: ZenodoDoi = "10.5072/zenodo.1199170"
-    epacems: ZenodoDoi = "10.5281/zenodo.8235497"
-    # epacems: ZenodoDoi = "10.5072/zenodo.1228519"
+    # epacems: ZenodoDoi = "10.5281/zenodo.8235497"
+    epacems: ZenodoDoi = "10.5072/zenodo.2237"
```
Comment (zaneselvans) [inline src/pudl/workspace/datastore.py:185]: Oooh, look it's mix-and-match production+sandbox DOIs and the number is sooo smol!

File: test/integration/epacems_test.py [hunk 2]
```diff
@@ -25,23 +25,23 @@ def epacems_parquet_path(
...
         pytest.skip("EPA CEMS not in settings file and so is not being tested.")
     path = epacems_parquet_path
-    years = epacems_year_and_state.years
-    # Use only Idaho if multiple states are given
-    states = (
-        epacems_year_and_state.states
-        if len(epacems_year_and_state.states) == 1
-        else ["ID"]
+    years = epacems_year_and_quarter.years
+    # Use only Idaho if multiple quarters are given
```
Comment (e-belfer) [inline test/integration/epacems_test.py:37]: 1st quarter, not Idaho

File: src/pudl/etl/epacems_assets.py [hunk 3]
```diff
@@ -64,23 +68,23 @@ def process_single_year(
     partitioned_path = PudlPaths().output_dir / "hourly_emissions_epacems"
     partitioned_path.mkdir(exist_ok=True)
 
-    for state in epacems_settings.states:
-        logger.info(f"Processing EPA CEMS hourly data for {year}-{state}")
-        df = pudl.extract.epacems.extract(year=year, state=state, ds=ds)
+    for quarter in epacems_settings.quarters:
+        logger.info(f"Processing EPA CEMS hourly data for {year}-{quarter}")
+        df = pudl.extract.epacems.extract(year=year, quarter=quarter, ds=ds)
         if not df.empty:  # If state-year combination has data
```
Comment (e-belfer) [inline src/pudl/etl/epacems_assets.py:74]: Should be quarter-year

File: src/pudl/etl/epacems_assets.py [hunk 4]
```diff
@@ -98,12 +102,21 @@ def consolidate_partitions(context, partitions: list[YearPartitions]) -> None:
     with pq.ParquetWriter(
         where=monolithic_path, schema=schema, compression="snappy", version="2.6"
     ) as monolithic_writer:
-        for year, states in partitions:
-            for state in states:
+        for year, quarters in partitions:
+            for state in EPACEMS_STATES:
                 monolithic_writer.write_table(
-                    pq.read_table(
```
Comment (e-belfer) [inline src/pudl/etl/epacems_assets.py:104]: "Concat a slice of each state's data" would be clearer

File: src/pudl/transform/epacems.py
Comment (cmgosnell) [inline src/pudl/transform/epacems.py:1]: update docs for year-quarter

**Review by zaneselvans — CHANGES_REQUESTED** [review]
I'm concerned about the way we're using year and quarter as independent partition variables, directly analogous to what we did with year and state (which were really independent) when there's only a single real dimension (time) that we're trying to describe. See comments.

File: test/integration/epacems_test.py [hunk 5]
```diff
@@ -25,23 +25,21 @@ def epacems_parquet_path(
...
-        else ["ID"]
-    )
+    years = epacems_year_and_quarter.years
+    states = ["ID"]  # Test on Idaho
     actual = epacems(
-        columns=["gross_load_mw"], epacems_path=path, years=years, states=states
+        columns=["gross_load_mw"],
+        epacems_path=path,
+        years=years,
+        states=states,
```
[the remark this answers is not in the document]
Comment (cmgosnell) [reply test/integration/epacems_test.py:42]: yes exactly. we've effectively preserved the old format of the table and thus all of the output tools

File: test/unit/settings_test.py [hunk 6]
```diff
@@ -91,34 +91,27 @@ def test_default_years(self: Self):
...
-            EpaCemsSettings(states=["fake_state"])
+            EpaCemsSettings(quarters=[5])
 
-    def test_duplicate_sort_states(self: Self):
-        """Test states are sorted and deduplicated."""
+    def test_duplicate_sort_quarters(self: Self):
+        """Test quarters are sorted and deduplicated."""
         with pytest.raises(ValidationError):
-            _ = EpaCemsSettings(states=["CA", "CA", "AL"])
+            _ = EpaCemsSettings(quarters=[4, 4, 2])
```
[the remark this answers is not in the document]
Comment (cmgosnell) [reply test/unit/settings_test.py:102]: we're no longer testing sorting so i made it clear that it is just duplication

File: src/pudl/etl/__init__.py [hunk 7]
```diff
@@ -88,6 +88,23 @@
...
+# By default, limit CEMS year processing concurrency to prevent memory overload.
+default_config = {
+    "execution": {
+        "config": {
+            "multiprocess": {
+                "tag_concurrency_limits": [
+                    {
+                        "key": "datasource",
+                        "value": "epacems",
+                        "limit": 2,
```
[the remark this answers is not in the document]
Comment (cmgosnell) [reply src/pudl/etl/__init__.py:100]: just the cems right now takes about 41 minutes

File: test/unit/workspace/datastore_test.py [hunk 8]
```diff
@@ -234,8 +234,8 @@ class TestZenodoFetcher(unittest.TestCase):
             },
         ]
     }
-    PROD_EPACEMS_DOI = "10.5281/zenodo.8235497"
-    PROD_EPACEMS_ZEN_ID = 8235497  # This is the last numeric part of doi
+    PROD_EPACEMS_DOI = "10.5281/zenodo.10233186"
+    PROD_EPACEMS_ZEN_ID = 10233186  # This is the last numeric part of doi
```
[the remark this answers is not in the document]
Comment (cmgosnell) [reply test/unit/workspace/datastore_test.py:238]: this has been updated via #3139

File: src/pudl/workspace/datastore.py [hunk 9]
```diff
@@ -194,7 +194,6 @@ class ZenodoFetcher:
     _descriptor_cache: dict[str, DatapackageDescriptor]
     zenodo_dois: ZenodoDoiSettings
     timeout: float
-    http: requests.Session
```
[the remark this answers is not in the document]
Comment (cmgosnell) [reply src/pudl/workspace/datastore.py:197]: I'm not sure. ella may know but this doesn't seem like a blocking issue.

[the remark this answers is not in the document]
Comment (zaneselvans) [reply src/pudl/workspace/datastore.py:197]: Yeah I think it's fine.

**Review by zaneselvans — APPROVED** [review]

**Comment by zaneselvans** [comment]
@cmgosnell

- I materialized the EPA CEMS assets locally so I could debug the test failures and it took nearly 2 hours, even with nothing else running.
- It's not immediately clear why it took so long though. It wasn't fully saturating either my CPUs or using all the memory.
- This is a very large performance regression. I think we need to put some effort into speeding this back up, or at least understanding why it's *so* slow. You said it "only" took 40 minutes for you, right?
- The integration / output tests were easy to fix and mostly had to do with changes to the `EpaCemsSettings` class, and the new expected behavior of the output function, which won't complain if you request state/year combos that don't exist. It just won't give you back the data it doesn't have.

**Comment by cmgosnell** [comment]
thanks for fixing that ci failure!

Its curious that it took you 2 hours. eeee. it has consistently taken my computer ~40 minutes. When it needed to download a new archive, it took 58. Which is still ~2x time from the previous setup. I 100% agree that we should take some time to make it faster, but I don't think we should delay integrating this before we do that.

**Comment by zaneselvans** [comment]
Another weird thing that I'm seeing locally comparing my two sets of outputs is the new ETL produces significantly smaller outputs, despite including a little bit more data.

```
4.8G	epacems_quarterly/
5.7G	hourly_emissions_epacems/
```
```
5.1G	hourly_emissions_epacems.parquet
5.6G	hourly_emissions_epacems.parquet (from nightly build)
```

**Comment by zaneselvans** [comment]
Hmm. Using the MacOS Activity Monitor (rather than btop) I see 10 python3.11 processes each of which claims to have 20 threads, and appears to be using multiple GB of memory, which means a lot of it is spilling onto swap on disk which would slow things down a lot.

**Pull request merged**